consensus/src/step_votes_reg.rs

```diff
@@ -94,33 +83,38 @@ impl CertificateInfo {
     fn is_ready(&self) -> bool {
         !self.cert.validation.is_empty()
             && !self.cert.ratification.is_empty()
-            && self.hash.is_some()
             && self.quorum_reached_validation
             && self.quorum_reached_ratification
     }
 
     /// Returns `true` if the certificate has empty hash
-    fn is_nil(&self) -> bool {
-        self.hash.map(|h| h == [0u8; 32]).unwrap_or_default()
+    fn is_nil_quorum(&self) -> bool {
+        self.hash == [0u8; 32]
+            && self.quorum_reached_ratification
+            && self.quorum_reached_validation
     }
 }
 
 pub type SafeCertificateInfoRegistry = Arc<Mutex<CertInfoRegistry>>;
 
+#[derive(Default, Clone)]
+struct CertInfo {
+    certs: HashMap<[u8; 32], CertificateInfo>,
```

Comment:
If the case is to allow only two options - Nil and Valid CertificateInfo, HashMap is a bit overkill here

Reply:
Agree with you

However, are we sure that no other hash can be filled there? I mean, if a provisioner send a third hash, is that message discarded before right?

Reply:
A third hash means either two candidates from the same generator, or there's a fork that reached the same height (and this should be handled separately)